File: benchmark/legendre/legendre.py
#!/bin/env pytriqs

import pytriqs.utility.mpi as mpi
from pytriqs.archive import HDFArchive
from pytriqs.operators import *
from pytriqs.applications.impurity_solvers.cthyb import *
from pytriqs.gf import *

spin_names = ("up","dn")
def mkind(spin): return (spin,0)

# Input parameters
beta = 10.0
U = 2.0
mu = 1.0
h = 0.0
V = 1.0
epsilon1 = 2.1
epsilon2 = -2.4

n_iw = 1025
n_tau = 10001
n_l = 50

p = {}
p["max_time"] = -1
p["random_name"] = ""
p["random_seed"] = 123 * mpi.rank + 567
p["length_cycle"] = 50
p["n_warmup_cycles"] = 50000
p["n_cycles"] = 5000000
p["measure_g_tau"] = False
p["measure_g_l"] = True

results_file_name = "legendre.h5"

mpi.report("Welcome to Legendre (1 correlated site + symmetric bath) test.")

H = U*n(*mkind("up"))*n(*mkind("dn"))

QN = []
for spin in spin_names: QN.append(n(*mkind(spin)))

gf_struct = {}
for spin in spin_names:
    bn, i = mkind(spin)
    gf_struct.setdefault(bn,[]).append(i)

mpi.report("Constructing the solver...")

# Construct the solver
S = SolverCore(beta=beta, gf_struct=gf_struct, n_tau=n_tau, n_iw=n_iw, n_l=n_l)

mpi.report("Preparing the hybridization function...")

# Set hybridization function    
delta_w = GfImFreq(indices = [0], beta=beta)
delta_w << (V**2) * inverse(iOmega_n - epsilon1) + (V**2) * inverse(iOmega_n - epsilon2)
for spin in spin_names:
    bn, i = mkind(spin)
    S.G0_iw[bn][i,i] << inverse(iOmega_n + mu - {'up':h,'dn':-h}[spin] - delta_w)

mpi.report("Running the simulation...")

# Solve the problem
S.solve(h_int=H, **p)

# Save the results
if mpi.is_master_node():
    with HDFArchive(results_file_name,'w') as Results:
        Results['G_l'] = S.G_l

File: test/c++/legendre.cpp
#include <triqs_cthyb/solver_core.hpp>

#include <triqs/operators/many_body_operator.hpp>
#include <triqs/hilbert_space/fundamental_operator_set.hpp>
#include <triqs/gfs.hpp>
#include <triqs/test_tools/gfs.hpp>

using namespace triqs_cthyb;
using triqs::operators::c;
using triqs::operators::c_dag;
using triqs::operators::n;
using namespace triqs::gfs;
using triqs::hilbert_space::gf_struct_t;

TEST(CtHyb, Legendre) {

  std::cout << "Welcome to the CTHYB solver\n";

  // Initialize mpi
  int rank = triqs::mpi::communicator().rank();

  // Parameters
  double beta     = 10.0;
  double U        = 2.0;
  double mu       = 1.0;
  double h        = 0.0;
  double V        = 1.0;
  double epsilon1 = 2.1;
  double epsilon2 = -2.4;

  // define operators and QN
  auto H = U * n("up", 0) * n("down", 0) + h * n("up", 0) - h * n("down", 0);
  // quantum numbers
  std::vector<many_body_op_t> qn{n("up", 0), n("down", 0)};
  // gf structure
  gf_struct_t gf_struct{{"up", {0}}, {"down", {0}}};

  // Construct CTQMC solver
  solver_core solver({beta, gf_struct, 1025, 2500, 50});

  // Set G0
  triqs::clef::placeholder<0> om_;
  auto g0_iw = gf<imfreq>{{beta, Fermion}, {1, 1}};
  g0_iw(om_) << om_ + mu - (V * V / (om_ - epsilon1) + V * V / (om_ - epsilon2));
  for (int bl = 0; bl < 2; ++bl) solver.G0_iw()[bl] = triqs::gfs::inverse(g0_iw);

  // Solve parameters
  int n_cycles       = 5000;
  auto p             = solve_parameters_t(H, n_cycles);
  p.random_name      = "";
  p.random_seed      = 123 * rank + 567;
  p.max_time         = -1;
  p.length_cycle     = 50;
  p.n_warmup_cycles  = 50;
  p.measure_G_tau    = false;
  p.measure_G_l      = true;
  p.quantum_numbers  = qn;
  p.partition_method = "quantum_numbers";
  p.move_double      = false;

  // Solve!
  solver.solve(p);

  auto &G_l = *solver.G_l;

  // Save the results
  if (rank == 0) {
    triqs::h5::file G_file("legendre.out.h5", 'w');
    h5_write(G_file, "G_up", G_l[0]);
    h5_write(G_file, "G_down", G_l[1]);
  }

  gf<legendre> g;
  if (rank == 0) {
    triqs::h5::file G_file("legendre.ref.h5", 'r');
    h5_read(G_file, "G_up", g);
    EXPECT_GF_NEAR(g, G_l[0]);
    h5_read(G_file, "G_down", g);
    EXPECT_GF_NEAR(g, G_l[1]);
  }
}
MAKE_MAIN;
